### ISSUE TITLE
Mapping from `RistrettoPoint` to and `EdwardsPoint` on the prime order subgroup

### ISSUE THREAD

**Issue by iquerejeta** [dalek-cryptography/curve25519-dalek#366]
tl;dr: let's define a function that maps a `RistrettoPoint` back to the `EdwardsPoint` structure, ensuring that the `EdwardsPoint` is part of the prime order subgroup. 

I know that the title is against the whole idea of the Ristretto technique, so I'll give a bit of context. 

In a nutshell, one of the cases where this would be useful, is to verify a Schnorr signature using the ed25519 verifier. This would allow to use the Schnorr compatible threshold/multi signatures (such as [FROST](https://github.com/isislovecruft/frost-dalek)) in an indistinguishable manner with respect to ed25519 signatures. In other words, there could be an ed25519 verifier that would be able to verify threshold/multi signatures without needing to change its verification logic, and without even knowing that it is verifying something different from an ed25519 (provided that there is an aggregator).

However, we have the problem of `let G be a group with prime order` for all these Schnorr compatible schemes---but we want to use it (or at least verify it) over a group with small cofactor. Ensuring that all the distributed protocol (for threshold/multi signature generation) happens in the prime order subgroup is cumbersome and might affect the security proofs (as nicely described [here](https://ristretto.group/why_ristretto.html)). The alternative that came to mind is to perform the whole threshold/multi signature generation mechanism using Ristretto, and once the public key and signature are generated, we map those points back to their Edwards representative in the prime order subgroup. This way we run the distributed protocol over a prime order group, but we still allow an ed25519 verifier to validate these signatures. 

`RistrettoPoint`s are internally represented by `EdwardsPoint`s. This `EdwardsPoint` is a canonical representative of an equivalence class. However, this "ristretto canonical representative" is not necessarily a point in the prime order subgroup, even if it is considered equivalent to one. So basically, what we want with the function discussed in this issue, is to extract that prime order point from the equivalence class defined by the `RistrettoPoint`. 

As far as I understood, to do so we would simply need to treat the `RistrettoPoint` as it's corresponding `EdwardsPoint`, and multiply it by the torsion-safe representative. 

Is this something you'd consider to include in the library? If so, I'd be happy to draft a first PR 👍

**Comment by str4d** [comment]
I'm strongly against this. The fact that `ristretto255` group elements happen to be represented internally by Edwards points in `curve25519-dalek`, is an internal implementation detail, and not something that any developer can rely on.

Quoting [draft-irtf-cfrg-ristretto255-decaf448-01](https://www.ietf.org/archive/id/draft-irtf-cfrg-ristretto255-decaf448-01.html#name-api-considerations) (disclosure: I am one of the authors of this ID):

> ## 6. API Considerations
>
> ristretto255 and decaf448 are abstractions which implement two prime-order groups, and their elements are represented by curve points, but they are not curve points.  The API needs to reflect that: the type representing an element of the group **SHOULD** be opaque and **MUST NOT** expose the underlying curve point or field elements.
>
> It is expected that a ristretto255 or decaf448 implementation can change its underlying curve without causing any breaking change. The ristretto255 and decaf448 constructions are carefully designed so that this will be the case, as long as implementations do not expose internal representations or operate on them except as described in this document.
>
> [...]
>
> ## 8. Security Considerations
>
> The ristretto255 and decaf448 groups provide higher-level protocols with the abstraction they expect: a prime-order group. Therefore, it's expected to be safer for use in any situation where Curve25519 or edwards448 is used to implement a protocol requiring a prime-order group. Note that the safety of the abstraction can be defeated by implementations that do not follow the guidance in Section 6.

While I am sympathetic to the constraints that existing deployments of Ed25519 are operating within, that is not sufficient justification to break open the Ristretto security boundary. Providing this kind of API would make it _far more likely_ that developers end up using Ristretto group elements in ways that are both unsafe, and potentially break their protocol entirely.

> This would allow to use the Schnorr compatible threshold/multi signatures (such as FROST) in an indistinguishable manner with respect to ed25519 signatures.

FROST is primarily being designed for the RedJubjub signature scheme, an instantiation of [RedDSA](https://zips.z.cash/protocol/protocol.pdf#concretereddsa) (which was designed to be similar to EdDSA, and supports cofactor curves) with an Edwards curve. It should therefore be possible to adapt FROST directly to Ed25519, as long as you are using the batch-validation-compatible Ed25519 signature validation equation (see [ZIP 215](https://zips.z.cash/zip-0215) for details on why this matters generally; here my reasoning is that RedDSA uses the same validation equation).

**Comment by iquerejeta** [comment]
Thanks for the reply. I understand that opening this type of function in the API is undesirable. In particular after seeing this, which I was not aware: 

> It is expected that a ristretto255 or decaf448 implementation can change its underlying curve without causing any breaking change.

which would certainly not be the case if we modify the API as discussed above. 

Indeed, it should be possible to adapt most schnorr-compatible multi/threshold signatures for cofactor groups (FROST was just an example, and not the particular case of interest), but that mechanism also opens some security concerns as we are modifying the protocol and need to make sure these modifications do not  affect the formal analysis of the scheme (one of the main motivations of Ristretto, if I understand correctly).

Regarding the verification equation, looking to validate it against latest `libsodium` version, so no luck there -.-

And given that I have some of the authors' attention, I might ask a quick question. Leaving the fact that this will not be part of the API, do you see any other concerns of this particular mechanism provided that: 

* the underlying curve of ristretto is curve25519
* the verification equation is the "strict" check provided by `libsodium`'s library

Thanks!

**Comment by isislovecruft** [comment]
Hi @iquerejeta. I'm also a fellow coauthor of the IETF Ristretto spec, and a coauthor of the IETF FROST draft. As @str4d has mentioned, we explicitly do not support mapping backwards from the Ristretto group to any underlying curve group. Verifiers which require implementations generic over group element types may already implement this by enforcing that the group element's type have methods for point compression, negation, equality, and (variable-time, double) scalar multiplication by the basepoint.
